File: 2914-find-the-safest-path-in-a-grid/find-the-safest-path-in-a-grid.cpp
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

class Solution {
public:
    int maximumSafenessFactor(vector<vector<int>>& grid) {
        int n = grid.size();
        
        // Edge Case: If start or end contains a thief, safeness factor is 0
        if (grid[0][0] == 1 || grid[n - 1][n - 1] == 1) {
            return 0;
        }
        
        // Step 1: Multi-Source BFS to compute minimum distance to any thief
        vector<vector<int>> dist(n, vector<int>(n, 1e9));
        queue<pair<int, int>> q;
        
        for (int r = 0; r < n; ++r) {
            for (int c = 0; c < n; ++c) {
                if (grid[r][c] == 1) {
                    dist[r][c] = 0;
                    q.push({r, c});
                }
            }
        }
        
        vector<pair<int, int>> directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        
        while (!q.empty()) {
            auto [r, c] = q.front();
            q.pop();
            
            for (auto& [dr, dc] : directions) {
                int nr = r + dr;
                int nc = c + dc;
                
                if (nr >= 0 && nr < n && nc >= 0 && nc < n && dist[nr][nc] == 1e9) {
                    dist[nr][nc] = dist[r][c] + 1;
                    q.push({nr, nc});
                }
            }
        }
        
        // Step 2: Modified Dijkstra using a Max-Heap
        // priority_queue in C++ acts as a Max-Heap by default
        priority_queue<vector<int>> pq; 
        pq.push({dist[0][0], 0, 0}); // stores: {safeness, r, c}
        
        vector<vector<int>> maxSafenessTo(n, vector<int>(n, -1));
        maxSafenessTo[0][0] = dist[0][0];
        
        while (!pq.empty()) {
            auto curr = pq.top();
            pq.pop();
            
            int currSafe = curr[0];
            int r = curr[1];
            int c = curr[2];
            
            if (r == n - 1 && c == n - 1) {
                return currSafe;
            }
            
            for (auto& [dr, dc] : directions) {
                int nr = r + dr;
                int nc = c + dc;
                
                if (nr >= 0 && nr < n && nc >= 0 && nc < n) {
                    int nextSafe = min(currSafe, dist[nr][nc]);
                    
                    if (nextSafe > maxSafenessTo[nr][nc]) {
                        maxSafenessTo[nr][nc] = nextSafe;
                        pq.push({nextSafe, nr, nc});
                    }
                }
            }
        }
        
        return 0;
    }
};

File: 2914-find-the-safest-path-in-a-grid/find-the-safest-path-in-a-grid.java
import java.util.*;

class Solution {
    public int maximumSafenessFactor(List<List<Integer>> grid) {
        int n = grid.size();
        
        // Edge Case: If start or end contains a thief, safeness factor is 0
        if (grid.get(0).get(0) == 1 || grid.get(n - 1).get(n - 1) == 1) {
            return 0;
        }
        
        // Step 1: Multi-Source BFS to compute minimum distance to any thief
        int[][] dist = new int[n][n];
        for (int[] row : dist) {
            Arrays.fill(row, Integer.MAX_VALUE);
        }
        
        Queue<int[]> queue = new LinkedList<>();
        for (int r = 0; r < n; r++) {
            for (int c = 0; c < n; c++) {
                if (grid.get(r).get(c) == 1) {
                    dist[r][c] = 0;
                    queue.offer(new int[]{r, c});
                }
            }
        }
        
        int[][] directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        
        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            int r = curr[0];
            int c = curr[1];
            
            for (int[] dir : directions) {
                int nr = r + dir[0];
                int nc = c + dir[1];
                
                if (nr >= 0 && nr < n && nc >= 0 && nc < n && dist[nr][nc] == Integer.MAX_VALUE) {
                    dist[nr][nc] = dist[r][c] + 1;
                    queue.offer(new int[]{nr, nc});
                }
            }
        }
        
        // Step 2: Modified Dijkstra using a Max-Heap
        // PriorityQueue sorts by safeness factor descending: (b[0] - a[0])
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> b[0] - a[0]);
        pq.offer(new int[]{dist[0][0], 0, 0});
        
        int[][] maxSafenessTo = new int[n][n];
        for (int[] row : maxSafenessTo) {
            Arrays.fill(row, -1);
        }
        maxSafenessTo[0][0] = dist[0][0];
        
        while (!pq.isEmpty()) {
            int[] curr = pq.poll();
            int currSafe = curr[0];
            int r = curr[1];
            int c = curr[2];
            
            if (r == n - 1 && c == n - 1) {
                return currSafe;
            }
            
            for (int[] dir : directions) {
                int nr = r + dir[0];
                int nc = c + dir[1];
                
                if (nr >= 0 && nr < n && nc >= 0 && nc < n) {
                    int nextSafe = Math.min(currSafe, dist[nr][nc]);
                    
                    if (nextSafe > maxSafenessTo[nr][nc]) {
                        maxSafenessTo[nr][nc] = nextSafe;
                        pq.offer(new int[]{nextSafe, nr, nc});
                    }
                }
            }
        }
        
        return 0;
    }
}